File: elixir/test/node_test.exs
defmodule Bonemesh.NodeTest do
  use ExUnit.Case, async: false

  alias Bonemesh.{Cert, Crypto, Node}

  @mesh "acme-prod"

  setup do
    {root_pub, root_priv} = Crypto.mldsa_generate(:mldsa87)
    %{root_pub: root_pub, root_priv: root_priv}
  end

  defp start_node(ctx, label) do
    {pub, priv} = Crypto.mldsa_generate(:mldsa65)
    now = System.system_time(:second)
    cert = Cert.new(@mesh, label, pub, now - 100, now + 3600) |> Cert.sign(ctx.root_priv)

    {:ok, node} =
      Node.start_link(
        label: label,
        mesh: @mesh,
        root_public: ctx.root_pub,
        cert: cert,
        id_public: pub,
        id_private: priv,
        port: 0
      )

    on_exit(fn -> if Process.alive?(node), do: Node.stop(node) end)
    node
  end

  defp await_data(match, timeout) do
    receive do
      {:bonemesh_data, payload} -> if match.(payload), do: :ok, else: await_data(match, timeout)
    after
      timeout -> :timeout
    end
  end

  test "two nodes exchange messages both directions", ctx do
    alpha = start_node(ctx, "alpha")
    beta = start_node(ctx, "beta")
    Node.add_listener(alpha, self())
    Node.add_listener(beta, self())

    {:ok, "beta"} = Node.connect(alpha, "127.0.0.1", Node.port(beta))

    assert Node.send(alpha, "beta", %{"m" => "ping"})
    assert :ok == await_data(&(&1["m"] == "ping"), 5000)

    assert Node.send(beta, "alpha", %{"m" => "pong"})
    assert :ok == await_data(&(&1["m"] == "pong"), 5000)
  end

  test "three node line relays across the middle hop", ctx do
    alpha = start_node(ctx, "alpha")
    beta = start_node(ctx, "beta")
    gamma = start_node(ctx, "gamma")
    Node.add_listener(gamma, self())

    {:ok, "beta"} = Node.connect(alpha, "127.0.0.1", Node.port(beta))
    {:ok, "beta"} = Node.connect(gamma, "127.0.0.1", Node.port(beta))

    # Wait for discovery to give alpha a route to gamma via beta.
    deadline = System.monotonic_time(:millisecond) + 15_000
    routed = wait_route(alpha, deadline)
    assert routed, "alpha never learned a route to gamma"
    assert :ok == await_data(&(&1["m"] == "relayed"), 5000)

    # The route to the non-neighbor gamma is via the middle hop beta — the
    # accessor the interop convergence tier reads.
    assert Node.routes(alpha)["gamma"] == "beta"
  end

  defp wait_route(alpha, deadline) do
    if Node.send(alpha, "gamma", %{"m" => "relayed"}) do
      true
    else
      if System.monotonic_time(:millisecond) < deadline do
        Process.sleep(200)
        wait_route(alpha, deadline)
      else
        false
      end
    end
  end

  test "large payload is chunked and reassembled end to end", ctx do
    alpha = start_node(ctx, "alpha")
    beta = start_node(ctx, "beta")
    Node.add_listener(beta, self())

    blob = String.duplicate("abcdefghij", 20_000) # 200 KB, well past the frame cap
    {:ok, "beta"} = Node.connect(alpha, "127.0.0.1", Node.port(beta))
    assert Node.send(alpha, "beta", %{"blob" => blob})
    assert :ok == await_data(&(&1["blob"] == blob), 10_000)
  end
end

File: js/test/node.test.js
// Node integration test: two real JS nodes over loopback TCP complete a BMX
// handshake and deliver an application payload in each direction. Proves the
// node wires handshake, transport, and framing together correctly; live
// cross-language delivery is proven by the interop matrix.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import crypto from 'node:crypto';
import { Node } from '../src/node.js';
import { build } from '../src/cert.js';
import { canonicalize } from '../src/canon.js';
import { mldsa65Generate } from '../src/crypto.js';

const MESH = 'acme-prod';

function newRoot() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync('ml-dsa-87');
  return { pubRaw: Buffer.from(publicKey.export({ format: 'jwk' }).pub, 'base64url'), privateKey };
}

function config(root, label) {
  const { pub, priv } = mldsa65Generate();
  const now = Math.floor(Date.now() / 1000);
  const cert = build(MESH, label, pub, now - 60, now + 3600);
  cert.sig = crypto.sign(null, Buffer.from(canonicalize(cert), 'utf8'), root.privateKey).toString('base64');
  return { label, mesh: MESH, rootPublic: root.pubRaw, cert, idPrivate: priv };
}

const waitFor = (predicate, timeoutMs = 5000) => new Promise((resolve, reject) => {
  const started = Date.now();
  const tick = () => {
    if (predicate()) return resolve();
    if (Date.now() - started > timeoutMs) return reject(new Error('timeout'));
    setTimeout(tick, 20);
  };
  tick();
});

test('two nodes exchange payloads over loopback', async () => {
  const root = newRoot();
  const one = await Node.start(config(root, 'one'), 0);
  const two = await Node.start(config(root, 'two'), 0);

  const oneGot = [];
  const twoGot = [];
  one.onMessage((p) => oneGot.push(p));
  two.onMessage((p) => twoGot.push(p));

  const peer = await two.connect('127.0.0.1', one.port());
  assert.equal(peer, 'one');

  assert.equal(two.send('one', { probe: 'to-one' }), true);
  await waitFor(() => oneGot.length > 0);
  assert.equal(oneGot[0].probe, 'to-one');

  // one replies to two over the same session (registered on accept).
  await waitFor(() => one.send('two', { probe: 'to-two' }));
  await waitFor(() => twoGot.length > 0);
  assert.equal(twoGot[0].probe, 'to-two');

  one.kill();
  two.kill();
});

test('three-node line relays across the middle hop', async () => {
  const root = newRoot();
  const alpha = await Node.start(config(root, 'alpha'), 0);
  const beta = await Node.start(config(root, 'beta'), 0);
  const gamma = await Node.start(config(root, 'gamma'), 0);

  const gammaGot = [];
  gamma.onMessage((p) => gammaGot.push(p));

  // Line topology: alpha <-> beta <-> gamma; alpha and gamma never connect.
  await alpha.connect('127.0.0.1', beta.port());
  await gamma.connect('127.0.0.1', beta.port());

  // Wait for discovery to give alpha a route to gamma (via beta), then delivery.
  await waitFor(() => alpha.send('gamma', { m: 'relayed' }), 15000);
  await waitFor(() => gammaGot.length > 0, 15000);
  assert.equal(gammaGot[0].m, 'relayed');
  assert.equal(alpha.routeTable().gamma, 'beta');

  alpha.kill();
  beta.kill();
  gamma.kill();
});
